pkg/grpc/labelstore/client/client.go

```diff
@@ -51,15 +59,31 @@ func (c Client) WatchMatches(selector klabels.Selector, labelType labels.Type, q
 		for {
 			labeled, err := watchClient.Recv()
 			if grpc.Code(err) == codes.Canceled {
+				c.logger.Infoln("label store client: terminating WatchMatches()")
 				// This just means quitCh fired and the RPC was canceled as expected
 				return
 			}
 
 			if err != nil {
-				// doesn't hurt to call cancelFunc() more than once potentially
-				cancelFunc()
-				c.logger.WithError(err).Errorln("Unexpected error reading from WatchMatches stream")
-				return
+				innerCancel()
+				c.logger.WithError(err).Errorln("unexpected error reading from WatchMatches stream, starting another RPC")
+
+				watchClient = nil
+
+				for watchClient == nil {
+					innerCtx, innerCancel = context.WithCancel(outerCtx)
+
+					time.Sleep(2 * time.Second)
```

Comment:
I'm conflicted about a global timeout. If such a timeout were to expire, the only action we can take to signal to the caller that things are broken is to close the output channel, at which point I would expect it to just reinitialize another watch. Do we think that would handle more different error cases?